crates/matrix-sdk/src/event_cache/pagination.rs

```diff
@@ -59,17 +59,75 @@ impl RoomPagination {
     /// This automatically takes care of waiting for a pagination token from
     /// sync, if we haven't done that before.
     ///
+    /// The `until` argument is an async closure that returns a [`ControlFlow`]
+    /// to decide whether a new pagination must be run or not. It's helpful when
+    /// the server replies with e.g. a certain set of events, but we would like
+    /// more, or the event we are looking for isn't part of this set: in this
+    /// case, `until` returns [`Control::Continue`], otherwise it returns
+    /// [`ControlFlow::Break`]. `until` receives [`BackPaginationOutcome`] as
+    /// its sole argument.
+    ///
     /// # Errors
     ///
     /// It may return an error if the pagination token used during
     /// back-pagination has disappeared while we started the pagination. In
     /// that case, it's desirable to call the method again.
-    #[instrument(skip(self))]
-    pub async fn run_backwards(&self, batch_size: u16) -> Result<BackPaginationOutcome> {
+    ///
+    /// # Example
+    ///
+    /// To do a single run:
+    ///
+    /// ```rust
+    /// use std::ops::ControlFlow;
+    ///
+    /// use matrix_sdk::event_cache::{
+    ///     BackPaginationOutcome,
+    ///     RoomPagination,
+    ///     TimelineHasBeenResetWhilePaginating
+    /// };
+    ///
+    /// # async fn foo(room_pagination: RoomPagination) {
+    /// let result = room_pagination.run_backwards(
+    ///     42,
+    ///     |BackPaginationOutcome { events, reached_start },
+    ///      _timeline_has_been_reset: TimelineHasBeenResetWhilePaginating| async move {
+    ///         // Do something with `events` and `reached_start` maybe?
+    ///         let _ = events;
+    ///         let _ = reached_start;
+    ///
+    ///         ControlFlow::Break(())
+    ///     }
+    /// ).await;
+    /// # }
+    #[instrument(skip(self, until))]
+    pub async fn run_backwards<Until, Break, UntilFuture>(
+        &self,
+        batch_size: u16,
+        mut until: Until,
+    ) -> Result<Break>
+    where
+        Until: FnMut(BackPaginationOutcome, TimelineHasBeenResetWhilePaginating) -> UntilFuture,
+        UntilFuture: Future<Output = ControlFlow<Break, ()>>,
+    {
+        let mut timeline_has_been_reset = TimelineHasBeenResetWhilePaginating::No;
+
         loop {
-            if let Some(result) = self.run_backwards_impl(batch_size).await? {
-                return Ok(result);
+            if let Some(outcome) = self.run_backwards_impl(batch_size).await? {
+                match until(outcome, timeline_has_been_reset).await {
+                    ControlFlow::Continue(()) => {
+                        debug!("back-pagination continues");
```

Comment:
Not sure this is useful to have, since there's a `debug!()` line already when the more exceptional timeline reset happens. If you really want to keep it, can we make it a trace log instead?

```suggestion
                        trace!("back-pagination continues");
```

Reply:
Let's remove it.